relay-server/src/envelope/container.rs

```diff
@@ -0,0 +1,422 @@
+{"items":[{"level":"info","message":"foobar"}]}
+        "#,
+        ))
+        .unwrap();
+
+        assert_eq!(item.item_count(), Some(1));
+        assert!(matches!(
+            ItemContainer::<TestLog>::parse(&item),
+            Err(ContainerParseError::MismatchedContentType {
+                expected: ContentType::LogContainer,
+                actual: None,
+            })
+        ));
+    }
+
+    #[test]
+    fn test_container_deserialize_missing_items() {
+        let (item, _) = Item::parse(Bytes::from_static(
+            br#"{"type":"log","content_type":"application/vnd.sentry.items.log+json","item_count":1}
+{"items2":[{"level":"info","message":"foobar"}]}
+        "#,
+        ))
+        .unwrap();
+
+        assert_eq!(item.item_count(), Some(1));
+        assert!(matches!(
+            ItemContainer::<TestLog>::parse(&item),
+            Err(ContainerParseError::Deserialize(_))
+        ));
+    }
+
+    #[test]
+    fn test_container_deserialize_unexpected_type() {
+        let (item, _) = Item::parse(Bytes::from_static(
+            br#"{"type":"log","content_type":"application/vnd.sentry.items.log+json","item_count":1}
+{"items":{"level":"info","message":"foobar"}}
+        "#,
+        ))
+        .unwrap();
+
+        assert_eq!(item.item_count(), Some(1));
+        assert!(matches!(
+            ItemContainer::<TestLog>::parse(&item),
+            Err(ContainerParseError::Deserialize(_))
+        ));
+    }
+
+    #[test]
+    fn test_container_deserialize_successful() {
+        let (item, _) = Item::parse(Bytes::from_static(
+            br#"{"type":"log","content_type":"application/vnd.sentry.items.log+json","item_count":2}
+{"items":[{"level":"info","message":"foobar"},{"level":"error","message":"ohno"}]}
+        "#,
+        ))
+        .unwrap();
+
+        assert_eq!(item.item_count(), Some(2));
+
+        let container = ItemContainer::<TestLog>::parse(&item).unwrap();
+        assert_debug_snapshot!(container, @r###"
+        ItemContainer {
+            items: [
+                TestLog {
+                    level: "info",
+                    message: "foobar",
+                },
+                TestLog {
+                    level: "error",
+                    message: "ohno",
+                },
+            ],
+        }
+        "###);
+    }
+
+    #[test]
+    fn test_container_roundtrip() {
+        let (item, _) = Item::parse(Bytes::from_static(
+            br#"{"type":"log","content_type":"application/vnd.sentry.items.log+json","item_count":2}
+{"items":[{"level":"info","message":"foobar"},{"level":"error","message":"ohno"}]}
+        "#,
+        ))
+        .unwrap();
+
+        let container = ItemContainer::<TestLog>::parse(&item).unwrap();
+        let mut new_item = Item::new(ItemType::Log);
+        container.write_to(&mut new_item).unwrap();
+
+        let container = ItemContainer::<TestLog>::parse(&new_item).unwrap();
+        assert_debug_snapshot!(container, @r###"
```

Comment:
Would it make sense here to assert that `item` and `new_item` are the same?

Reply:
Tricky, because of missing impls, do you think the snapshot doesn't cover enough?

Reply:
No, it's fine the way it is.